depdive/src/update.rs

```diff
@@ -92,6 +96,41 @@ pub enum VersionConflict {
     },
 }
 
+#[derive(Debug, Clone)]
+pub struct FileUnsafeChangeStats {
+    pub file: String,
+    pub change_type: Delta,
+    pub unsafe_delta: UnsafeDelta, // Delta in Unsafe counter:
+    // doesn't verify if unsafe code has not been touched
```

Comment:
Thanks for the example. In that case, the most conservative way to think is that all changes in a crate containing unsafe can potentially introduce safety problem? Lesser conservative is to detect if any files containing unsafe code is present in the diff which is done in this PR. Another way forward, could be to see if files containing multiple modules, and perform the checking on a modular level.

Reply:
All changes to a **module** containing unsafe could change safety guarantees.

So for example, a diff removing the last bit of unsafe code from a module is a very strong signal of safety improvement. A diff to safe lines of code in a module containing unsafe is unknown and would need inspection. A diff to unsafe lines of code in a module is probably a reasonable signal that safety changes need inspection. A diff adding the first instance of unsafe to a module is a strong signal of decreased safety.

I'm not sure there is a way to judge which of the two middle options is worse. But I believe the way you were calculating this, you only surfaced the 3rd kind, which is diff deltas including unsafe blocks.

Reply:
While its probably fairly difficult to do at the module level (since that would require being able to actually parse rust files) we could at least do an approximation of that and just check if a file containing unsafe was changed (since 90% of the time a file contains a single module)

Reply:
@metajack thoughts on this approach?

Reply:
> just check if a file containing unsafe was changed (since 90% of the time a file contains a single module)
this is what is being done in this PR.

Reply:
Got it, then surfacing the following metrics should be sufficient: 1) unsafe line count delta and 2) list of files which contain unsafe which have changes between versions.

Reply:
[this is ](https://github.com/nasifimtiazohi/whackadep/blob/67f678087213bbd807176c1811a9f797c39745cd/depdive/src/update.rs#L86) is what is being done as well. It returns the `list of files which contain unsafe which have changes between versions` along with their `unsafe line count delta`. Do you think naming is confusing?

Reply:
Those both sound good. Maybe `unsafe_line_delta` and `unsafe_file_changed`? Or something like that?

Reply:
changed the latter. `unsafe_delta` is better as it is not about lines, but in terms of functtions, exprs, impls etc as given by geiger counter.